File: src/simulation/culture/group/AspectStratum.java
package simulation.culture.group;

import simulation.culture.aspect.Aspect;
import simulation.culture.aspect.AspectController;
import simulation.culture.aspect.AspectResult;
import simulation.culture.aspect.ConverseWrapper;
import simulation.culture.group.centers.Group;
import simulation.culture.group.centers.PopulationCenter;
import simulation.culture.group.request.EvaluatorsKt;
import simulation.culture.thinking.meaning.ConstructMemeKt;
import simulation.culture.thinking.meaning.MemePool;
import simulation.space.Territory;
import simulation.space.resource.tag.ResourceTag;
import simulation.culture.aspect.dependency.Dependency;
import simulation.culture.group.request.ResourceEvaluator;
import simulation.culture.thinking.meaning.Meme;
import simulation.space.resource.MutableResourcePack;

import java.util.*;

import static simulation.Controller.session;

public class AspectStratum implements Stratum {
    private int population;
    /**
     * How many people have already worked on this turn;
     */
    private int workedAmount = 0;
    private boolean isRaisedAmount = false;
    private List<ConverseWrapper> aspects = new ArrayList<>();
    private Map<ResourceTag, MutableResourcePack> dependencies = new HashMap<>();
    private List<Meme> popularMemes = new ArrayList<>();

    public AspectStratum(int population, ConverseWrapper aspect) {
        this.population = population;
        aspects.add(aspect);
        aspect.getDependencies().getMap().keySet().forEach(tag -> {
            if (tag.isInstrumental() && !tag.name.equals("phony")) {
                dependencies.put(tag.copy(), new MutableResourcePack());
            }
        });
    }

    public int getWorkedAmount() {
        return workedAmount;
    }

    public MutableResourcePack getInstrumentByTag(ResourceTag tag) {
        MutableResourcePack resourcePack = dependencies.get(tag);
        return resourcePack == null ? new MutableResourcePack() : resourcePack;
    }

    public int getPopulation() {
        return population;
    }

    public int getFreePopulation() {
        return population - workedAmount;
    }

    public List<ConverseWrapper> getAspects() {
        return aspects;
    }

    public boolean containsAspect(ConverseWrapper aspect) {
        return aspects.contains(aspect);
    }

    public void decreaseAmount(int delta) {
        population -= delta;
        if (workedAmount > population) {
            workedAmount = population;
        }
    }

    public void decreaseWorkedAmount(int delta) {
        workedAmount -= delta;
    }

    public void useAmount(int amount) {
        if (amount <= 0) {
            return;
        }
        this.workedAmount += amount;
        if (workedAmount > population) {
            this.population = workedAmount;
            isRaisedAmount = true;
        }
    }

    public void update(
            MutableResourcePack accessibleResources,
            Territory accessibleTerritory,
            Group group
    ) {
        if (population == 0) {
            return;
        }
        int oldPopulation = population;
        MutableResourcePack pack = use(new AspectController(
                1,
                getFreePopulation(),
                getFreePopulation(),
                EvaluatorsKt.getPassingEvaluator(),
                group.getPopulationCenter(),
                accessibleTerritory,
                false,
                null,
                group
        ));
        population = oldPopulation;
        accessibleResources.addAll(pack);
        updateTools(accessibleTerritory, group);
        isRaisedAmount = false;
    }

    public MutableResourcePack use(AspectController controller) {
        MutableResourcePack resourcePack = new MutableResourcePack();
        for (Aspect aspect : aspects) {
            AspectResult result = aspect.use(controller);
            if (result.resources.isNotEmpty()) {
                popularMemes.add(ConstructMemeKt.constructMeme(aspect));
                result.resources.getResources().forEach(r -> popularMemes.add(ConstructMemeKt.constructMeme(r)));
            }
            if (result.isFinished) {
                resourcePack.addAll(result.resources);
            }
            result.pushNeeds(controller.getGroup());
        }
        return resourcePack;
    }

    private void updateTools(Territory accessibleTerritory, Group group) {
        if (!session.isTime(session.stratumTurnsBeforeInstrumentRenewal)) {
            return;
        }
        for (Map.Entry<ResourceTag, MutableResourcePack> entry : dependencies.entrySet()) {
            int currentAmount = entry.getValue().getAmount();
            if (currentAmount >= population) {
                continue;
            }
            if (!entry.getKey().isInstrumental()) {
                continue;
            }
            ResourceEvaluator evaluator = EvaluatorsKt.tagEvaluator(entry.getKey());
            for (Aspect aspect : aspects) {//TODO choose the best
                if (currentAmount >= population) {
                    break;
                }
                Set<Dependency> deps = aspect.getDependencies().getMap().get(entry.getKey());
                if (deps != null) {
                    for (Dependency dependency : deps) {
                        AspectResult result = dependency.useDependency(
                                new AspectController(
                                        1,
                                        population - currentAmount,
                                        1,
                                        evaluator,
                                        group.getPopulationCenter(),
                                        accessibleTerritory,
                                        false,
                                        null,
                                        group
                                ));
                        if (result.isFinished) {
                            currentAmount += evaluator.evaluate(result.resources);
                            entry.getValue().addAll(evaluator.pick(result.resources));//TODO disband
                            if (currentAmount >= population) {
                                break;
                            }
                        }
                    }
                }
            }
        }
    }

    public void finishUpdate(Group group) {
        popularMemes.forEach(m -> group.getCultureCenter().getMemePool().strengthenMeme(m));
        popularMemes.clear();
        if (workedAmount < population) {
            population = workedAmount;
        }
        if (aspects.get(0).getName().equals("TakeOnHerbivore")) {
            int i = 0;
        }
        workedAmount = 0;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        AspectStratum stratum = (AspectStratum) o;
        return Objects.equals(aspects, stratum.aspects);
    }

    @Override
    public int hashCode() {
        return Objects.hash(aspects);
    }

    @Override
    public String toString() {
        StringBuilder stringBuilder = new StringBuilder("Stratum with population ");
        stringBuilder.append(population).append(", aspects: ");
        for (Aspect aspect : aspects) {
            stringBuilder.append(aspect.getName()).append(" ");
        }
        return stringBuilder.toString();
    }
}

File: src/shmp/simulation/culture/group/stratum/AspectStratum.kt
package shmp.simulation.culture.group.stratum

import shmp.random.singleton.chanceOfNot
import shmp.simulation.CulturesController.session
import shmp.simulation.culture.aspect.AspectController
import shmp.simulation.culture.aspect.ConverseWrapper
import shmp.simulation.culture.aspect.MeaningInserter
import shmp.simulation.culture.aspect.dependency.Dependency
import shmp.simulation.culture.aspect.getAspectImprovement
import shmp.simulation.culture.group.centers.Group
import shmp.simulation.culture.group.passingReward
import shmp.simulation.culture.group.request.*
import shmp.simulation.culture.thinking.meaning.Meme
import shmp.simulation.culture.thinking.meaning.makeMeme
import shmp.simulation.space.territory.Territory
import shmp.simulation.space.resource.container.MutableResourcePack
import shmp.simulation.space.resource.tag.ResourceTag
import shmp.simulation.space.tile.Tile
import java.util.*
import kotlin.math.ceil
import kotlin.math.max
import kotlin.math.min

class AspectStratum(
        override var population: Int,
        val aspect: ConverseWrapper,
        tile: Tile
) : BaseStratum(tile, "Stratum of aspect ${aspect.name}", "") {
    private var _effectiveness = -1.0

    private var workedAmount = 0
    private var isRaisedAmount = false
    private val dependencies: MutableMap<ResourceTag, MutableResourcePack> = HashMap()
    private val popularMemes: MutableList<Meme> = ArrayList()
    override var importance: Int
        get() = aspect.usefulness
        set(value) {
            aspect.gainUsefulness(value - importance)
        }

    fun getInstrumentByTag(tag: ResourceTag) = dependencies[tag] ?: MutableResourcePack()

    override val freePopulation: Int
        get() = population - workedAmount
    override val cumulativeWorkAblePopulation: Double
        get() = freePopulation * effectiveness

    init {
        aspect.dependencies.map.keys.forEach { tag: ResourceTag ->
            if (tag.isInstrumental && tag.name != "phony")
                dependencies[tag.copy()] = MutableResourcePack()
        }
    }

    override fun decreaseAmount(amount: Int) {
        population -= amount
        if (workedAmount > population)
            workedAmount = population
    }

    fun decreaseWorkedAmount(amount: Int) {
        workedAmount -= amount
    }

    private val effectiveness: Double
        get() {
            if (_effectiveness == -1.0) {
                _effectiveness = 1.0 +
                        places
                                .flatMap { it.owned.resources }
                                .map { it.getAspectImprovement(aspect) }
                                .foldRight(0.0, Double::plus)
            }
            return _effectiveness
        }

    override fun useAmount(amount: Int, maxOverhead: Int): StratumPeople {
        if (amount <= 0)
            return StratumPeople(0, this)

        var actualAmount = ceil(amount / effectiveness).toInt()
        if (actualAmount <= freePopulation)
            workedAmount += actualAmount
        else {
            val additional = min(maxOverhead, actualAmount - freePopulation)
            actualAmount = additional + freePopulation
            population += additional
            workedAmount = population
            if (additional > 0)
                isRaisedAmount = true
        }

        return StratumPeople(actualAmount, this, effectiveness)
    }

    override fun update(accessibleResources: MutableResourcePack, accessibleTerritory: Territory, group: Group) {
        super.update(accessibleResources, accessibleTerritory, group)
        if (population == 0)
            return

        if (aspect !is MeaningInserter) {
            val oldPopulation = population
            val evaluator = passingEvaluator
            val overhead = aspect.calculateNeededWorkers(evaluator, freePopulation.toDouble()).toDouble()
            val amount = aspect.calculateProducedValue(evaluator, freePopulation)
            val pack = use(AspectController(
                    1,
                    amount,
                    amount,
                    evaluator,
                    group.populationCenter,
                    accessibleTerritory,
                    false,
                    group,
                    group.cultureCenter.meaning
            ))

            if (population < oldPopulation)
                population = oldPopulation
            accessibleResources.addAll(pack)
        }

        updateInfrastructure(accessibleTerritory, group)
        isRaisedAmount = false
        ego.update(accessibleResources, accessibleTerritory, group, this)
    }

    fun use(controller: AspectController): MutableResourcePack {
        val resourcePack = MutableResourcePack()
        val result = aspect.use(controller)

        if (result.resources.isNotEmpty) {
            popularMemes.add(makeMeme(aspect))
            result.resources.resources.forEach { popularMemes.add(makeMeme(it)) }
        }
        if (result.isFinished)
            resourcePack.addAll(result.resources)
        result.pushNeeds(controller.group)

        return resourcePack
    }

    private fun updateInfrastructure(accessibleTerritory: Territory, group: Group) {
        session.stratumInstrumentRenewalProb.chanceOfNot {
            return
        }

        if (population == 0)
            return

        updateTools(accessibleTerritory, group)
        updatePlaces(group)
    }

    private fun updateTools(accessibleTerritory: Territory, group: Group) {
        for ((key, value) in dependencies) {
            var currentAmount = value.amount
            if (currentAmount >= population)
                continue
            if (!key.isInstrumental)
                continue

            val evaluator = tagEvaluator(key)
            //TODO choose the best
            if (currentAmount >= population)
                break
            val deps: Set<Dependency>? = aspect.dependencies.map[key]
            if (deps != null)
                for (dependency in deps) {
                    val result = dependency.useDependency(
                            AspectController(
                                    1,
                                    (population - currentAmount).toDouble(),
                                    1.0,
                                    evaluator,
                                    group.populationCenter,
                                    accessibleTerritory,
                                    false,
                                    group,
                                    group.cultureCenter.meaning
                            ))
                    if (result.isFinished) {
                        currentAmount += (evaluator.evaluate(result.resources)).toInt()
                        value.addAll(evaluator.pickAndRemove(result.resources))
                        group.populationCenter.turnResources.addAll(result.resources)

                        if (currentAmount >= population)
                            break
                    }
                }
        }
    }

    private fun updatePlaces(group: Group) {
        if (!group.territoryCenter.settled)
            return

        val request: Request = AspectImprovementRequest(
                aspect,
                RequestCore(
                        group,
                        0.5,
                        0.5 * max(1, importance),
                        passingReward,
                        passingReward,
                        30,
                        setOf(RequestType.Improvement)
                )
        )
        val (pack, usedAspects) = group.populationCenter.executeRequest(request)

        usedAspects.forEach {
            it.gainUsefulness((2 / session.stratumInstrumentRenewalProb).toInt())
        }
        pack.resources.forEach { addEnhancement(it, group) }
    }

    override fun finishUpdate(group: Group) {
        popularMemes.forEach { group.cultureCenter.memePool.strengthenMeme(it) }
        popularMemes.clear()

        if (workedAmount < population && !ego.isActive)
            population = workedAmount

        workedAmount = 0
        _effectiveness = -1.0
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val stratum = other as AspectStratum
        return aspect == stratum.aspect
    }

    override fun hashCode() = Objects.hash(aspect)

    override fun toString() = "Stratum with population $population, " +
            "effectiveness $effectiveness, importance $importance, ${aspect.name}, Places: " +
            innerPlaces.joinToString { "$it " } +
            super.toString()

    override fun die() {
        population = 0
        workedAmount = 0
        innerPlaces.forEach { session.world.strayPlacesManager.addPlace(it) }
    }
}